### PR TITLE
Refactor dev

### LINKED ISSUE #20: Device `blas::axpy` does not have `const` modifier on `x`
Hello all,

The API for `blas::axpy` on the CPU is:
```
void axpy(
    int64_t n,
    double alpha,
    double const *x, int64_t incx,
    double       *y, int64_t incy )
```
with a const on `x`.

However, the device version has API:
```
void blas::axpy(
    int64_t n,
    double alpha,
    double *dx, int64_t incdx,
    double *dy, int64_t incdy,
    blas::Queue& queue)
```
which lacks the `const` modifier on `dx`.

Would it be possible to add it? I would prefer to have symmetry in my device and CPU implementation. Is there concerns that `dx` may be modified?

FYI, this work is part of @BallisticLA. I'm not sure where that falls into the ECP.

### PR DESCRIPTION
**Pull request by mgates3**
Refactors GPU device routines in similar fashion to PR #18 for CPU routines.
1. Low-level overloaded wrappers call cublas, rocblas, or onemkl (in `blas::internal` namespace).
2. Mid-level templated wrapper checks and converts arguments, calls low-level wrapper (in `blas::impl` namespace).
3. High-level overloaded wrappers call mid-level wrapper (in `blas` namespace).

This also fixes #20.

I also fixed a few minor deprecation & compilation issues in example_gemm.cc and cblas_wrappers.cc.

For most files, rather than looking at the "Files changed" diff, I suggest comparing the existing CPU implementation with this updated GPU implementation like this (bash syntax):
```
# Fetch this branch from my repo. The `-f` is needed only if re-fetching an updated version.
git fetch -f https://github.com/mgates3/blaspp.git refactor-dev:refactor-dev
git checkout refactor-dev

# Copy CPU BLAS 1, 2, 3 files to tmp directory.
mkdir tmp
cp src/asum.cc src/axpy.cc src/copy.cc src/dot.cc src/iamax.cc src/nrm2.cc src/scal.cc src/swap.cc src/gemv.cc src/ger.cc src/hemv.cc src/her.cc src/her2.cc src/symv.cc src/syr.cc src/syr2.cc src/trmv.cc src/trsv.cc src/gemm.cc src/hemm.cc src/her2k.cc src/herk.cc src/symm.cc src/syr2k.cc src/syrk.cc src/trmm.cc src/trsm.cc tmp/

# Rename files with "device_" prefix and apply simple code transformations.
cd tmp
for x in *.cc ; do mv $x device_$x ; done

perl -pi -e 's/(\b|to_)blas_int/${1}device_blas_int/g;  s/.*(op|uplo|diag|side)2char.*\n//;  s@/// CPU@/// GPU device@;'  *.cc
cd ..

# Apply simple code transformations to low-level wrappers.
perl -pe 's/cublas([SDCZ])/rocblas_\l$1/; s/cublas/rocblas/g; s/CUBLAS/ROCBLAS/g; s/cuComplex/rocblas_float_complex/g; s/cuDoubleComplex/rocblas_double_complex/g; s/Batched\(/_batched(/;' src/cublas_wrappers.cc > tmp/rocblas_wrappers.cc

perl -pe 's/cublas([SDCZ])(\w+)\(/oneapi::mkl::blas::$2(/; s/queue.handle\(\)/dev_queue/; s/cublas/onemkl/g; s/\((const )?cu(Double)?Complex\*\) *//g; s/CUBLAS/ONEMKL/g;' src/cublas_wrappers.cc > tmp/onemkl_wrappers.cc

# Compare directories. Under meld's "File Filters", I uncheck "New" to ignore new files.
# Could also use Xcode's opendiff. In that case, check "Added/Deleted" to ignore new files.
meld tmp src
```
For the most part, the only differences are adding `queue` and setting the device.

In the low-level wrappers, cuBLAS and rocBLAS are very similar, while oneMKL has a few more differences, including slightly different spacing that I'm not going to worry about right now.

Don't worry about the stub_wrappers.cc because it is removed in a subsequent PR.

### CODE REVIEW

File: src/device_batch_gemm.cc [hunk 1]
```diff
@@ -92,20 +92,20 @@ void gemm(
...
-                device::batch_sgemm( queue,
-                                    transB_, transA_,
-                                    n_, m_, k_,
-                                    alpha[0], dBarray, lddb_, dAarray, ldda_,
-                                    beta[0],  dCarray, lddc_,
-                                    ibatch);
+                internal::batch_gemm(
+                    transB_, transA_,
+                    n_, m_, k_,
+                    alpha[0], dBarray, lddb_, dAarray, ldda_,
```
Comment (dbielich) [inline src/device_batch_gemm.cc:98]: same as other comments: missed changing from lddb to ldb, ldda to lda and lddc to ldc

File: src/device_batch_trsm.cc [hunk 2]
```diff
@@ -85,11 +85,11 @@ void blas::batch::trsm(
...
 
-            device::batch_strsm( queue,
-                                side_, uplo_, trans_, diag_,
-                                m_, n_, alpha[0],
-                                dAarray, ldda_,
-                                dBarray, lddb_, ibatch);
+            internal::batch_trsm(
+                side_, uplo_, trans_, diag_,
+                m_, n_, alpha[0],
+                dAarray, ldda_,
```
Comment (dbielich) [inline src/device_batch_trsm.cc:91]: ld*d*a/b

Comment (dbielich) [reply src/device_batch_trsm.cc:91]: and below

File: src/stub_wrappers.cc [hunk 3]
```diff
@@ -3,901 +3,839 @@
...
-void saxpy(
-    blas::Queue& queue,
+//------------------------------------------------------------------------------
+void axpy(
     device_blas_int n,
     float alpha,
-    float *dx, device_blas_int incdx,
-    float *dy, device_blas_int incdy)
+    float const* dx, device_blas_int incdx,
+    float *dy, device_blas_int incdy,
```
Comment (dbielich) [inline src/stub_wrappers.cc:23]: remove the d in incdx/y

Comment (dbielich) [reply src/stub_wrappers.cc:23]: same for other instances in this file

File: src/stub_wrappers.cc [hunk 4]
```diff
@@ -3,901 +3,839 @@
...
-// -----------------------------------------------------------------------------
-// sgemm
-void sgemm(
-    blas::Queue& queue,
+//------------------------------------------------------------------------------
+void gemm(
     blas::Op transA, blas::Op transB,
     device_blas_int m, device_blas_int n, device_blas_int k,
     float alpha,
     float const *dA, device_blas_int ldda,
```
Comment (dbielich) [inline src/stub_wrappers.cc:313]: additionally the ldda, lddb, lddc for consistency.

File: src/rocblas_wrappers.cc [hunk 5]
```diff
@@ -82,14 +81,13 @@ void saxpy(
...
-// daxpy
-void daxpy(
-    blas::Queue& queue,
+//------------------------------------------------------------------------------
+void axpy(
     device_blas_int n,
     double alpha,
-    double *dx, device_blas_int incdx,
-    double *dy, device_blas_int incdy)
+    double const* dx, device_blas_int incdx,
```
Comment (dbielich) [inline src/rocblas_wrappers.cc:88]: inc*d*x here and inc*d*y

File: src/rocblas_wrappers.cc [hunk 6]
```diff
@@ -133,126 +129,118 @@ void zaxpy(
...
-         queue.handle(),
-         n,
-         (const rocblas_float_complex*) dx, incdx,
-         (const rocblas_float_complex*) dy, incdy,
-         (rocblas_float_complex*) result));
+    blas_dev_call(
+        rocblas_cdotu(
+            queue.handle(),
+            n,
+            (const rocblas_float_complex*) dx, incdx,
```
Comment (dbielich) [inline src/rocblas_wrappers.cc:214]: maybe change instances of dx/dy to x/y

File: src/onemkl_wrappers.cc [hunk 7]
```diff
@@ -85,14 +84,13 @@ void saxpy(
...
-// daxpy
-void daxpy(
-    blas::Queue& queue,
+//------------------------------------------------------------------------------
+void axpy(
     device_blas_int n,
     double alpha,
-    double *dx, device_blas_int incdx,
-    double *dy, device_blas_int incdy)
+    double const* dx, device_blas_int incdx,
```
Comment (dbielich) [inline src/onemkl_wrappers.cc:91]: inc*d*x and inc*d*y

File: src/onemkl_wrappers.cc [hunk 8]
```diff
@@ -61,19 +61,18 @@ oneapi::mkl::side side2onemkl(blas::Side side)
...
-// saxpy
-void saxpy(
-    blas::Queue& queue,
+//------------------------------------------------------------------------------
+void axpy(
     device_blas_int n,
     float alpha,
-    float *dx, device_blas_int incdx,
-    float *dy, device_blas_int incdy)
+    float const* dx, device_blas_int incdx,
```
Comment (dbielich) [inline src/onemkl_wrappers.cc:73]: *d*x and *d*y change without *d*

File: src/device_internal.hh [hunk 9]
```diff
@@ -11,578 +11,588 @@
...
-// axpy
-void saxpy(
-    blas::Queue& queue,
+//------------------------------------------------------------------------------
+void axpy(
     device_blas_int n,
     float alpha,
-    float *dx, device_blas_int incdx,
-    float *dy, device_blas_int incdy);
+    float const* dx, device_blas_int incdx,
```
Comment (dbielich) [inline src/device_internal.hh:47]: *d*x, *d*y, inc*d*x, inc*d*y

File: src/device_batch_gemm.cc (hunk 1 above)
Comment (mgates3) [reply src/device_batch_gemm.cc:98]: All these for batch routines fixed in PR #22.

File: src/stub_wrappers.cc (hunk 3 above)
Comment (mgates3) [reply src/stub_wrappers.cc:23]: I didn't change these in the wrappers.cc files yet. That could be a later PR. Here I just made the minimal changes needed to fix `const`. The stub_wrappers.cc file gets deleted later.

**Review by dbielich — APPROVED** [review]

**Pull request merged**